xdsl/transforms/constant_fold_interp.py

```diff
@@ -80,7 +79,7 @@ class ConstantFoldInterpPass(ModulePass):
     name = "constant-fold-interp"
 
     def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
-        interpreter = Interpreter(op)
+        interpreter = Interpreter(op, index_bitwidth=32)
```

Comment:
I'm not sure that this is a sound decision? Constant folding index values on 64 bit platforms will yield wrong results if I'm not mistaken?

Reply:
Yep, as discussed privately first, I would default to using platform.architecture to figure out a sound host index size

Reply:
> I'm not sure that this is a sound decision? Constant folding index values on 64 bit platforms will yield wrong results if I'm not mistaken?

Why would it?

Reply:
Hmm, interesting question @Moxinilian, I dug a bit into [the docs](https://mlir.llvm.org/docs/Dialects/IndexOps/) and the provided example there (bottom of the dialect introduction section) helps to construct an example:

Assuming we try to fold `div(0x100000002, 2)` (or any other operation that right shifts)

- if we are 64 bit, but simulate 32, we truncate then divide, giving us `0x01` as a result
- if we are 32 bit, but simulate 64, we get an upper bit set that would have been truncated, giving us `0x80000001` as a result

MLIR always uses 64 bit values for constant folding `index`, and solved this conundrum by not being able to fold right shifting operations on indices. If we are able to pull information through into the folding pass, we will be allowed to actually do such things.

Reply:
isn't index lowering also parametrized by the target's index size?

Reply:
As far as I can see, the `canonicalize` pass has no index bitwidth argument, and that pass does folding if I'm not mistaken.